### ISSUE TITLE
Future of @deprecated methods

### ISSUE THREAD

**Issue by msangel** [bkiers/Liqp#254]
I gave some minutes and took a closer look at what we have now. Central concept: There is a new `TemplateParser` class with `Template parse([File|String|InputStream|Reader] input){...}` methods. And all the parse methods on `Template` are marked as deprecated, so only `render(...)` methods should live there. 
This is good as follows the [SRP](https://en.wikipedia.org/wiki/Single-responsibility_principle). Now one class is about to parse content into a [immutable] template and another is about to render it.
None of the methods were removed so far due to backward compatibility. BUT I think this is anyway huge change and it makes sense to release simply a new major version where will be no deprecated methods, just new logic.

Also, things to fix(removing `deprecated` will do the job, just enumerating notable here): 
- protectionSettings, renderSettings and parseSettings should live in one place. As proposed by @kohlschuetter that place is `TemplateParser` and all the using places (Template and its TemplateContext) simply have ref to their owning `TemplateParser`. 
- Remove methods on `Insertion` hierarchy that was about adding new insertions that have been migrated to the new Insertion**s** class(by @kohlschuetter).
- Remove all the `Template` constructors that ware for creating template from the template source input. Now for this we have `TemplateParser`.
- Remove that `Filter.apply(...)` overloaded method version that is not taking `TemplateContext` as the first parameter (that has been marked as deprecated for a while) 

Some propositions to discuss:
- protectionSettings, renderSettings and parseSettings are just containers where we put and read values from. Since now we have `TemplateParser` why simply not remove them and put those values directly into `TemplateParser`? The interface will be simpler: instead of `new TemplateParser.Builder().withRenderSettings(new RenderSettings.Builder().withStrictVariables(true)).build()` simply have: `new TemplateParser.Builder().withStrictVariables(true).build()`

In conclusion: all of the mutable logic was moved to `TemplateParser.Builder`, `TemplateParser` itself just consumes the template source by creating `Template` and the `Template` is immutable and safe and does only rendering. 

Note: I was for keeping existing static methods on `Template` class and I am still for it, pretty sure we can reimplement old behavior even with these changes.

@bkiers  @kohlschuetter

**Comment by kohlschuetter** [comment]
Removing the deprecated methods will certainly break backwards compatibility. I don't know enough about existing customers, but besides being a mostly stable product, it looks like there aren't many public projects depending on Liqp. I think that may be justification enough to not worry too much about removing these methods and simply go ahead and remove them for the next "major" release.

However, if the goal is to maintain backwards compatibility with the static methods in `Template`, there may not be too much to cut.

I think this conversation could be a good opportunity to think ahead and do some more, breaking changes to the API (i.e., give the next release some more thought and time).

For example:
- `Template` right now is a class. It could be an interface instead. Depending on the `TemplateParser` configuration, different implementations could be chosen.
- `Template.render` currently always returns strings. I've looked into ways of not having to compose a big concatenated `String` if all we need to do is to stream the rendered output to a `Writer`. For that, I implemented `Template.prerender`, which returns objects that can be turned in to strings on demand (or directly streamed to `Writer` without much of concatenation, allowing early-abort on strings that become too large, support lazy evaluation of included templates, etc.). I will submit a pull request for that soon. I use `prerender` in combination with my "stringhold" library (https://github.com/kohlschutter/stringhold) to reduce the amount of string concatenations/reallocations.

Regarding the ParseSettings/RenderSettings/ProtectionSettings: I think these are a good way to group settings by concern. If we expect a lot more tune-able parameters, having them all in one place may become too crowded. However, I don't have strong feelings about the situation right now. From a performance standpoint, the separation doesn't seem to matter. From a code readability standpoint, I wished Java would provide a more compact way of writing these configurations, but it's not something I feel as urgent as the other refactoring issues.

**Comment by bkiers** [comment]
> [...] release simply a new major version where will be no deprecated methods
> [...] why simply not remove them and put those values directly into `TemplateParser`? 

I have no objections :+1: 

> Removing the deprecated methods will certainly break backwards compatibility.

Which would be OK when we go from `0.x` to `1.x`. Existing users that do not want to refactor their code into using the new API are free to keep using the old version.

> I think this conversation could be a good opportunity to think ahead and do some more, breaking changes to the API (i.e., give the next release some more thought and time)
> 
> * [...]

Sound like decent additions. But given I am no longer actively using/developing Liqp much, I'll let @msangel decide if/when/how a new major release will be done.

**Comment by allantodd** [comment]
I've just upgraded and removed the deprecated methods.  The problem I faced is that I'm working on a spring application and I had circular dependencies.  I have it working but it's quite possible I'm doing something wrong so I'd be interested in your advice as the docs still show the old way of doing things.

I have a couple of tags, the simplest is one to translate days of the week to other languages.  I use it like this

`"Heute ist {% dayofweek dayToTranslate %}"`

where `dayToTranslate` is a parameter holding a day name such as 'Monday'

The render function basically looks like this (I've simplified a little).  We're using Kotlin

```
    override fun render(context: TemplateContext, vararg nodes: LNode): Any {
        val dayParam = super.asString(nodes[0].render(context), context)
        val englishDayName = templateParserSupplier().parse("{{$dayParam}}").render(context.variables)
        val languageCode = context.variables[MESSAGE_PARAM_LANGUAGE] as String?
        return translateEnglishDayName(englishDayName, languageCode)
    }
```

The problem I had is that to build the TemplateParser I need the Tags, but to create the Tags I need the TemplateParser.

I worked around this using `templateParserSupplier()`

The constructor of the Tag class looks like this

```
open class DayOfWeekTag(
    private val dayOfWeekTranslator: DayOfWeekTranslator,
    private val templateParserSupplier: () -> TemplateParser,
) : Tag("dayofweek") {
...
}
```

I created a new Spring config class to deal with registering the tags and building the `TemplateParser` as follows.

```
@Configuration
open class LiqpConfig : ApplicationContextAware {

    private lateinit var applicationContext: ApplicationContext

    private val templateParser: TemplateParser by lazy {
        val parseSettings = ParseSettings.Builder().apply {
            val insertions = applicationContext.getBeansOfType(Insertion::class.java).values
            insertions.forEach {
                with(it)
            }
            val filters = applicationContext.getBeansOfType(Filter::class.java).values
            filters.forEach {
                with(it)
            }
        }.build()
        TemplateParser.Builder().withParseSettings(parseSettings).build()
    }

    override fun setApplicationContext(applicationContext: ApplicationContext) {
        this.applicationContext = applicationContext
    }

    @Bean
    open fun templateParserSupplier(): () -> TemplateParser = { templateParser }
}
```

It's working but it seems difficult and wasn't required before.  I have the feeling I'm missing a trick.  Maybe using the TemplateParser in the Tag render method is an unusual thing to do, it seemed quite natural when I wrote it a few years ago.

Thank you

**Comment by allantodd** [comment]
That branch works okay for me, thanks.  I'm building the TemplateParser like this (in Kotlin)

```
    private val templateParser: TemplateParser by lazy {
        val parseSettings = ParseSettings.Builder().apply {
            val insertions = applicationContext.getBeansOfType(Insertion::class.java).values
            insertions.forEach {
                with(it)
            }
            val filters = applicationContext.getBeansOfType(Filter::class.java).values
            filters.forEach {
                with(it)
            }
        }.build()
        TemplateParser.Builder().withParseSettings(parseSettings).build()
    }
```

**Comment by allantodd** [comment]
As mentioned above though, it does seem to create this chicken and egg problem that you need the tags to create the TemplateParser, but the tags need to use the TemplateParser (or mine do at least).  

I appreciate that the tags don't need the parser at the point they are created, only when they are used.  However if you're using dependency injection you can't inject each into the other without resolving the circular dependency with a workaround.

**Comment by msangel** [comment]
Closed as resolved. New issues should raise new reports.
